### PR TITLE
[Port] AI Upgrades

### PR DESCRIPTION
**Pull request by imsynthex**
## About The Pull Request

Ports:
- [x] https://github.com/yogstation13/Yogstation/pull/12815
- [x] https://github.com/yogstation13/Yogstation/pull/12919
- [x] https://github.com/yogstation13/Yogstation/pull/12989
- [x] https://github.com/yogstation13/Yogstation/pull/12999
- [x] https://github.com/yogstation13/Yogstation/pull/13051

TODO:
- [x] Remove Cloud Resources tab from the dashboard
- [x] Split Med and Diagnostics HUDs into smaller seperate projects
- [x] Remove the firewall project (no AI management console)
- [x] Remove AI Data Core remnants from dashboard
- [x] Add CPU and RAM boards to AI research and sci lathe

## Why It's Good For The Game

Makes AI gameplay slightly more sophisticated, than "Shift click doors for 2 hours"

## Changelog
:cl: TheGamerdk, ImSynthex
add: The AI can now download upgrades, these are accessible through the "Processing Dashboard" button on the HUD, the upgrades range from accessing Security HUDs to being able to examine people through cameras
add: Added CPU and RAM expansion cards for Science to build, these expansion cards allow the AI download upgrades faster in the case of CPU and store more upgrades in the case of RAM
/:cl:

### CODE REVIEW

File: code/modules/mob/living/silicon/ai/decentralized/projects/ai_huds.dm [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/datum/ai_project/security_hud
+	name = "Advanced Security HUD"
+	description = "Using experimental long range passive sensors should allow you to detect various implants such as loyalty implants and tracking implants."
```
Comment (AnCopper) [inline code/modules/mob/living/silicon/ai/decentralized/projects/ai_huds.dm]:
```suggestion
	description = "Using experimental long range passive sensors should allow you to detect various implants such as mind shields and tracking implants."
```

They're called mind shields on this server.

**Comment by imsynthex** [comment]
Alright, checks are failing because every map has an old AI core in it, gonna have to get to that later

**Comment by Archanial** [comment]
Please credit original authors in changelog.

**Comment by zeskorion** [comment]
Do not port this, ai doesnt need to be powergame job

**Comment by imsynthex** [comment]
> Do not port this, ai doesnt need to be powergame job

I genuinely don't get what what you mean by this making AI a "powergame job", do mean the upgrades are powergaming? Do you think the ability transfer between data cores is powergaming? Please specify so I can actually maybe fix it

**Comment by DatBoiTim** [comment]
> 

I think powercrepe is more the correct word.

Also it makes AI infinitely harder to kill, and generally makes any deal of going against them a pain.

**Comment by imsynthex** [comment]
Yea, I agree it's a bit on the OP end with AI being much more difficult to kill, I'll try and find a good way to balance it, currently AI suffers permadeath if all data cores are gone, which is sort of better than nothing

**Comment by Bibby0110** [comment]
Might be a good idea to search through Yogstations github for some of the additional PR's. Took a quick look and some seems to be missing (see https://github.com/yogstation13/Yogstation/pull/13013, probably others)

**Comment by Bibby0110** [comment]
> Also it makes AI infinitely harder to kill, and generally makes any deal of going against them a pain.

Oh yes it very much ups the skill-ceiling of what the AI is able to do to stay alive. Definitely needs some dialing back

**Comment by ghost** [comment]
Wouldn't this makes the AI harder to kill? The AI would know you're trying to kill him the second you kill 1 Core. and you have to evade through validhunters, security and secborgs to reach the other **7** data cores all around the whole station just to kill him. It's also less cool if you think about it. Busting through the AI satellite ala Payday seems more cool.

**Comment by imsynthex** [comment]
> Wouldn't this makes the AI harder to kill? The AI would know you're trying to kill him the second you kill 1 Core. and you have to evade through validhunters, security and secborgs to reach the other **7** data cores all around the whole station just to kill him. It's also less cool if you think about it. Busting through the AI satellite ala Payday seems more cool.

Yeah, that's why pretty much the only viable strategy rn is to subvert it rather than trying to kill it, will have to think of a way to balance it

**Comment by Bibby0110** [comment]
> Wouldn't this makes the AI harder to kill? The AI would know you're trying to kill him the second you kill 1 Core. and you have to evade through validhunters, security and secborgs to reach the other **7** data cores all around the whole station just to kill him. It's also less cool if you think about it. Busting through the AI satellite ala Payday seems more cool.

If the AI manages to get 7 cooled rooms setup to do this, yes. The easiest ways would in this case be subversion, coordinated bombs, power sink (for making *some* of them go offline), or attaching signalers to disable the others cores, while going guns blazing ala Payday style in and killing the core in the satellite.

**Comment by DatBoiTim** [comment]
> If the AI manages to get 7 cooled rooms setup to do this, yes.

It's not mentioned the operating parameters, nor how much heat they create.

**Comment by Bibby0110** [comment]
> > If the AI manages to get 7 cooled rooms setup to do this, yes.
> 
> It's not mentioned the operating parameters, nor how much heat they create.

Oh, right. The description hasn't been copied over. Feel free to read the original PR to get the details
EDIT: It's not mentioned there either, ha!
They don't work above 17C or in a vacuum. They produce approx. 1,875C to 2.5C per tick

**Comment by DatBoiTim** [comment]
> > > If the AI manages to get 7 cooled rooms setup to do this, yes.
> > 
> > 
> > It's not mentioned the operating parameters, nor how much heat they create.
> 
> Oh, right. The description hasn't been copied over. Feel free to read the original PR to get the details EDIT: It's not mentioned there either, ha! They don't work above 17C or in a vacuum.

Ok but that means nothing if they don't generate a decent amount of heat

**Comment by Bibby0110** [comment]
> > > > If the AI manages to get 7 cooled rooms setup to do this, yes.
> > > 
> > > 
> > > It's not mentioned the operating parameters, nor how much heat they create.
> > 
> > 
> > Oh, right. The description hasn't been copied over. Feel free to read the original PR to get the details EDIT: It's not mentioned there either, ha! They don't work above 17C or in a vacuum.
> 
> Ok but that means nothing if they don't generate a decent amount of heat

1,875C to 2.5C per tick. Enough to require a cooler setup if you don't want it going offline in a few minutes (Depending on heat capacity and room-size, which again, require more effort to change up)

**Comment by ghost** [comment]
Also makes it extremely easy to destroy the AI by carding/subverting it instead Just steal the password from RD's office and print AI upload module along with intellicard/law change module removing all the fun from the heist.

**Comment by Bibby0110** [comment]
> Also makes it extremely easy to destroy the AI by carding/subverting it instead Just steal the password from RD's office and print AI upload module along with intellicard/law change module removing all the fun from the heist.

That is something that's a bit of a hassle to get over yeah. If you're able to steal the password and some access you'd be able to remotely download the AI. With the caveat that the AI would be notified that you're attempting to downloading it, initially from which area, and later from the exact coords. So if you're able to hold off security for a while it's the "safe" option. (AI also has the possibility to prolong the download time by using the upgrade system)

**Comment by MCterra10** [comment]
Not a fan. Imo the current AI system is pretty well balanced and doesn't really need a complete overhaul that will introduce heaps of bugs and glaring balance issues.

**Comment by MCterra10** [comment]
God imagine antag RD with steal AI objective. Literally easier to do than stealing a plasma tank loll

**Comment by MCterra10** [comment]
this comment on the original PR raises an interesting point too:
clockies
Bee has reworked clock cult as a valid gamemode. How is AI conversion going to work with this PR?

**Comment by AnCopper** [comment]
> God imagine antag RD with steal AI objective. Literally easier to do than stealing a plasma tank loll

Heads can't be antags on Bee.

**Comment by SynergyJohn** [comment]
I think that this is a decent change and I support it, at least to see if the community likes it because it adds more of an intricacy than >spam thermite >spam PKA in between turret blasts >EMP

**Comment by ghost** [comment]
> I think that this is a decent change and I support it, at least to see if the community likes it because it adds more of an intricacy than >spam thermite >spam PKA in between turret blasts >EMP

I mean breaking into RD office. looking at the password. Making a upload board. subvert the seems even more simple than breaking into the sattelite

**Comment by Bibby0110** [comment]
> > I think that this is a decent change and I support it, at least to see if the community likes it because it adds more of an intricacy than >spam thermite >spam PKA in between turret blasts >EMP
> 
> I mean breaking into RD office. looking at the password. Making a upload board. subvert the seems even more simple than breaking into the sattelite

This doesn’t change anything about law upload. The password in the RDs office is only for uploading and downloading AIs, not laws.

Idk if you guys have changed anything about law uploads but they’re still stupidly easy where I come from, and after subverting an AI it’s still easy to just waltz in and card the AI after

**Comment by Hibabo** [comment]
> > I think that this is a decent change and I support it, at least to see if the community likes it because it adds more of an intricacy than >spam thermite >spam PKA in between turret blasts >EMP
> 
> I mean breaking into RD office. looking at the password. Making a upload board. subvert the seems even more simple than breaking into the sattelite

wont the password be on a piece of paper like the decryption monitor key? what stops the RD from just taking it round start and either securing it somewhere else or keeping it on him.

**Comment by Bibby0110** [comment]
> > > I think that this is a decent change and I support it, at least to see if the community likes it because it adds more of an intricacy than >spam thermite >spam PKA in between turret blasts >EMP
> > 
> > 
> > I mean breaking into RD office. looking at the password. Making a upload board. subvert the seems even more simple than breaking into the sattelite
> 
> wont the password be on a piece of paper like the decryption monitor key? what stops the RD from just taking it round start and either securing it somewhere else or keeping it on him.

It’s printed on a plaque in the office. You have to stand next to the plaque to read the password off of it. 
I don’t know your rules but I’m sure going out of your way to destroy that plaque would be against them in some capacity?

@ImSynthex also the “ When a data core is deleted, a /mob/living/silicon/ai spawns in it's place” isn’t actually a bug per-say, it’s just invisible to everyone but ghosts. It’s so we can let the AI get its death message before we delete it. Technically you could move it somewhere like null-space if you wanted.

**Comment by imsynthex** [comment]
> wont the password be on a piece of paper like the decryption monitor key?

The current solution is having the password on a plaque in the RD's office, I'd be happy to make it a piece of paper in the RD's locker though, as to make it another de facto high risk item like the spare AA safe codes

> the “ When a data core is deleted, a /mob/living/silicon/ai spawns in it's place” isn’t actually a bug per-say, it’s just invisible to everyone but ghosts. It’s so we can let the AI get its death message before we delete it. Technically you could move it somewhere like null-space if you wanted.

Oh, I'll keep that in mind then

**Comment by VictorVow** [comment]
I'm not a fan of AI Data Cores (which just give the AI extra lives), but I do like the idea of Expansion Card Busses machines which can be used to upgrade an AIs capabilities. 

AI Cores being singular is balanced around the existence of the AI satellite for protecting it. If multiple AI Data Cores were possible, then the AI Sat should be deleted and the starting Core should be on-station e.g. in the Science Server room or AI Upload room. This is because having the option of more Cores dotted around the station would be equally as difficult to sabotage compared to 1 Core in the satellite. 

Having 1 Core in the satellite PLUS the possibility of more in highly secure places like the Armory or Vault means that the AI would be many times harder to kill.

**Comment by imsynthex** [comment]
> AI Cores being singular is balanced around the existence of the AI satellite for protecting it. If multiple AI Data Cores were possible, then the AI Sat should be deleted and the starting Core should be on-station e.g. in the Science Server room or AI Upload room. This is because having the option of more Cores dotted around the station would be equally as difficult to sabotage compared to 1 Core in the satellite.

I gave some thought to this and I agree, it seems like a good middle ground solution would be to keep the current AI core and add expansion card buses to not mess too much with the balance while still improving AI gameplay

I'll see about doing that once I get some more time to work on this

**Comment by Bibby0110** [comment]
> > AI Cores being singular is balanced around the existence of the AI satellite for protecting it. If multiple AI Data Cores were possible, then the AI Sat should be deleted and the starting Core should be on-station e.g. in the Science Server room or AI Upload room. This is because having the option of more Cores dotted around the station would be equally as difficult to sabotage compared to 1 Core in the satellite.
> 
> I gave some thought to this and I agree, it seems like a good middle ground solution would be to keep the current AI core and add expansion card buses to not mess too much with the balance while still improving AI gameplay
> 
> I'll see about doing that once I get some more time to work on this

Thankfully separating the two shouldn’t be too difficult, don’t hesitate to contact me if you need help with that

**Comment by imsynthex** [comment]
Alright, all the features are done for this PR, no doubt there'll be a million oversights by me so this will need a testmerge

I can also make the wiki changes if/when this gets merged

**Comment by zeskorion** [comment]
Wait, i seem to recall yog removing borgs as well tho

File: code/modules/mob/living/carbon/human/examine.dm [hunk 2]
```diff
@@ -375,3 +375,117 @@
...
+
+		msg += "<B>[capitalize(t_his)] [parse_zone(t)] is missing!</B>\n"
+
+	if(l_limbs_missing >= 2 && r_limbs_missing == 0)
+		msg += "[t_He] look[p_s()] all right now.\n"
+	else if(l_limbs_missing == 0 && r_limbs_missing >= 2)
+		msg += "[t_He] really keeps to the left.\n"
+	else if(l_limbs_missing >= 2 && r_limbs_missing >= 2)
+		msg += "[t_He] [p_do()]n't seem all there.\n"
+
```
Comment (Archanial) [inline code/modules/mob/living/carbon/human/examine.dm:483]: Pretty sure this doesn't handle better limbs 
@Kapu1178

File: code/modules/mob/living/silicon/ai/upgrades/management/ai_dashboard.dm [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+		qdel(src)
+	owner = new_owner
+	available_projects = list()
+	completed_projects = list()
+	running_projects = list()
+	cpu_usage = list()
+	ram_usage = list()
+
+	for(var/path in subtypesof(/datum/ai_project))
+		available_projects += new path(owner, src)
```
Comment (Archanial) [inline code/modules/mob/living/silicon/ai/upgrades/management/ai_dashboard.dm:25]: How often does this run?

File: code/modules/mob/living/carbon/human/examine.dm (hunk 2 above)
Comment (Kapu1178) [reply code/modules/mob/living/carbon/human/examine.dm:483]: Me no like. This completely undoes the changes I made to examine for Kapulimbs. Also, not knowing exactly what limbs someone is missing is **annoying**

**Comment by imsynthex** [comment]
I didn't plan to keep working on this so I'll close it until I feel like working on it again, which might be in a long while

**Pull request closed without merging**